core/src/main/scala/io/finch/Accept.scala

```diff
@@ -1,26 +1,72 @@
 package io.finch
 
-import javax.activation.MimeType
-import scala.util.control.NonFatal
+import java.util.Locale
+
+import shapeless.Witness
 
 /**
  * Models an HTTP Accept header (see RFC2616, 14.1).
  *
+ * @note This API doesn't validate the input primary/sub types.
+ *
  * @see https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html
  */
-final case class Accept(primary: String, sub: String) {
-  def matches(that: Accept): Boolean =
-    (this.primary == that.primary || this.primary == "*") &&
-      (this.sub == that.sub || this.sub == "*")
+abstract class Accept {
+  def primary: String
+  def sub: String
+  def matches[CT <: String](implicit m: Accept.Matcher[CT]): Boolean = m(this)
+
+  override def toString: String = s"Accept: $primary/$sub"
 }
 
 object Accept {
 
+  private object Empty extends Accept {
+    def primary: String = ""
+    def sub: String = ""
+    override def matches[CT <: String](implicit m: Matcher[CT]): Boolean = false
+  }
+
+  abstract class Matcher[CT <: String] {
+    def apply(a: Accept): Boolean
+  }
+
+  object Matcher {
+
+    private object Empty extends Matcher[Nothing] {
+      def apply(a: Accept): Boolean = false
+    }
+
+    implicit val json: Matcher[Application.Json] = fromWitness[Application.Json]
+    implicit val xml: Matcher[Application.Xml] = fromWitness[Application.Xml]
+    implicit val text: Matcher[Text.Plain] = fromWitness[Text.Plain]
+    implicit val html: Matcher[Text.Html] = fromWitness[Text.Html]
+
+    implicit def fromWitness[CT <: String](implicit w: Witness.Aux[CT]): Matcher[CT] = {
+      val slashIndex = w.value.indexOf(47)
+      if (slashIndex == 0 || slashIndex == w.value.length) Empty.asInstanceOf[Matcher[CT]]
+      else new Matcher[CT] {
+        private val primary: String = w.value.substring(0, slashIndex).trim.toLowerCase(Locale.ENGLISH)
+        private val sub: String = w.value.substring(slashIndex + 1, w.value.length).trim.toLowerCase(Locale.ENGLISH)
+        def apply(a: Accept): Boolean =
+          (a.primary == primary || a.primary == "*") && (a.sub == sub || a.sub == "*")
```

Comment:
Validation accepts a wrong case - `Accept(*, not_star_value)`. It seems like a minor problem but may become complicated due to the lack of logging inside Finch.

Suggestion - `(a.primary == "*" && a.sub == "*") || (a.primary == primary && (a.sub == sub || a.sub == "*"))` .

Reply:
Sounds good to me! @ImLiar thoughts?

Reply:
Yes, it's reasonable